File: THCS/2022_2023/LongAn_2223/Bai1/Bai1.py
def sum_of_divisors(n):
    """Hàm tính tổng các ước số của n."""
    total = 0
    for i in range(1, int(n**0.5) + 1):
        if n % i == 0:
            total += i
            if i != n // i:
                total += n // i
    return total

def find_friendly_number(N):
    """Tìm số M lớn hơn N cùng tỉ ước."""
    sum_N = sum_of_divisors(N)
    ratio_N = sum_N / N

    for M in range(N + 1, 100001):
        sum_M = sum_of_divisors(M)
        ratio_M = sum_M / M

        if ratio_N == ratio_M:
            return M  # Trả về số M tìm được

    return -1  # Nếu không tìm thấy số nào thỏa mãn

if __name__ == "__main__":
    N = int(input("Nhập vào số N: "))
    M = find_friendly_number(N)
    print(M)

File: THCS/2022_2023/LongAn_2223/Bai1/Bai1.cpp
#include <iostream>
#include <vector>

using namespace std;

// Hàm tính tổng các ước số của số n
int sumOfDivisors(int n) {
    int sum = 0;
    for (int i = 1; i * i <= n; ++i) {
        if (n % i == 0) {
            sum += i; // i là ước
            if (i != n / i) {
                sum += n / i; // n / i cũng là ước
            }
        }
    }
    return sum;
}

// Hàm tìm số M lớn hơn N cùng tỉ ước
int findFriendlyNumber(int N) {
    int sumN = sumOfDivisors(N);
    double ratioN = static_cast<double>(sumN) / N;

    for (int M = N + 1; M <= 100000; ++M) {
        int sumM = sumOfDivisors(M);
        double ratioM = static_cast<double>(sumM) / M;

        if (ratioN == ratioM) {
            return M; // Trả về số M tìm được
        }
    }

    return -1; // Nếu không tìm thấy số nào thỏa mãn
}

int main() {
    int N;
    cout << "Nhap vao so N: ";
    cin >> N;

    int M = findFriendlyNumber(N);
    cout << M << endl;

    return 0;
}

